src/main/java/net/minecraftforge/common/capabilities/CapabilityProvider.java

```diff
@@ -5,206 +5,139 @@
 
 package net.minecraftforge.common.capabilities;
 
+import java.util.function.Supplier;
+
 import javax.annotation.ParametersAreNonnullByDefault;
 
-import com.google.common.annotations.VisibleForTesting;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
 import net.minecraft.MethodsReturnNonnullByDefault;
-import net.minecraft.nbt.CompoundTag;
 import net.minecraft.core.Direction;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.world.level.chunk.LevelChunk;
-import net.minecraftforge.common.util.LazyOptional;
 import net.minecraftforge.event.ForgeEventFactory;
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
-
-import java.util.function.Supplier;
 
+/**
+ * The default implementation of {@link ICapabilityProvider}.
+ *
+ * @param <T> The type of the owner object.
+ */
 @MethodsReturnNonnullByDefault
 @ParametersAreNonnullByDefault
-public abstract class CapabilityProvider<B extends ICapabilityProviderImpl<B>> implements ICapabilityProviderImpl<B>
+public abstract class CapabilityProvider<T extends ICapabilityProvider> implements ICapabilityProvider
 {
-    @VisibleForTesting
-    static boolean SUPPORTS_LAZY_CAPABILITIES = true;
 
-    private final @NotNull Class<B> baseClass;
-    private @Nullable CapabilityDispatcher capabilities;
-    private boolean valid = true;
+    @Nullable
+    private CapabilityDispatcher<T> attachedCaps = null;
+    protected boolean capsValid = true;
+    protected CompoundTag lazyCapNbt = new CompoundTag();
```

Comment:
I think this can be made `null` since you're already checking for `!= null` below. It's also one less object allocation.